### ISSUE TITLE
AttributeError: 'NoneType' object has no attribute 'rsplit'

### ISSUE THREAD

**Issue by jjkoehorst** [ResearchObject/runcrate#33]
Trying another workflow with newly transformed expressionTool to commandLineTool...
Unfortunately this zip file is 109MB. If you need a copy let me know how I can share this with you.

```
/Volumes/Git/m-unlock/cwl/tests/PROV: sha256 manifest lists snapshot/array_to_file_tool.cwl multiple times with the same value
/Volumes/Git/m-unlock/cwl/tests/PROV: sha256 manifest lists snapshot/concatenate.cwl multiple times with the same value
/Volumes/Git/m-unlock/cwl/tests/PROV: sha1 manifest lists snapshot/array_to_file_tool.cwl multiple times with the same value
/Volumes/Git/m-unlock/cwl/tests/PROV: sha1 manifest lists snapshot/concatenate.cwl multiple times with the same value
/Volumes/Git/m-unlock/cwl/tests/PROV: sha512 manifest lists snapshot/array_to_file_tool.cwl multiple times with the same value
/Volumes/Git/m-unlock/cwl/tests/PROV: sha512 manifest lists snapshot/concatenate.cwl multiple times with the same value
Traceback (most recent call last):
  File "/Users/jasperk/mambaforge/bin/runcrate", line 8, in <module>
    sys.exit(cli())
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/cli.py", line 68, in convert
    crate = builder.build()
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/convert.py", line 266, in build
    self.add_workflow(crate)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/convert.py", line 322, in add_workflow
    self.add_step(crate, workflow, s)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/convert.py", line 334, in add_step
    tool = self.add_tool(crate, workflow, cwl_step.run)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/convert.py", line 380, in add_tool
    self.add_param_connections(crate, tool)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/convert.py", line 623, in add_param_connections
    from_param = get_fragment(s)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/convert.py", line 115, in get_fragment
    return uri.rsplit("#", 1)[-1]
AttributeError: 'NoneType' object has no attribute 'rsplit'
```

**Comment by jjkoehorst** [comment]
I am working on recreating the files see #36 #35 and now I have a zip file with the cwl files but it cannot find it...

http://download.systemsbiology.nl/unlock/cwl/issues/PROV_missing_cwl_but_it_is_there.zip

Should I create a new issue for this one?

```
/Volumes/Git/m-unlock/cwl/tests/PROV: sha512 manifest lists snapshot/array_to_file_tool.cwl multiple times with the same value
/Volumes/Git/m-unlock/cwl/tests/PROV: sha512 manifest lists snapshot/concatenate.cwl multiple times with the same value
/Volumes/Git/m-unlock/cwl/tests/PROV: sha256 manifest lists snapshot/array_to_file_tool.cwl multiple times with the same value
/Volumes/Git/m-unlock/cwl/tests/PROV: sha256 manifest lists snapshot/concatenate.cwl multiple times with the same value
/Volumes/Git/m-unlock/cwl/tests/PROV: sha1 manifest lists snapshot/array_to_file_tool.cwl multiple times with the same value
/Volumes/Git/m-unlock/cwl/tests/PROV: sha1 manifest lists snapshot/concatenate.cwl multiple times with the same value
Traceback (most recent call last):
  File "/Users/jasperk/mambaforge/bin/runcrate", line 8, in <module>
    sys.exit(cli())
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/cli.py", line 68, in convert
    crate = builder.build()
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/convert.py", line 266, in build
    self.add_workflow(crate)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/convert.py", line 322, in add_workflow
    self.add_step(crate, workflow, s)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/convert.py", line 334, in add_step
    tool = self.add_tool(crate, workflow, cwl_step.run)
  File "/Users/jasperk/mambaforge/lib/python3.10/site-packages/runcrate/convert.py", line 345, in add_tool
    cwl_tool = self.cwl_defs[tool_fragment]
KeyError: 'fastp.cwl'
```

**Comment by simleo** [comment]
> Should I create a new issue for this one?

If it's a different error on a different RO bundle, better create a new issue. Always link to the specific RO bundle file so it can be associated to the issue

**Comment by jjkoehorst** [comment]
Correct zip file http://download.systemsbiology.nl/unlock/cwl/issues/PROV_rsplit.zip

**Comment by simleo** [comment]
Recap of the situation on http://download.systemsbiology.nl/unlock/cwl/issues/PROV_rsplit.zip:
 * The original `AttributeError: 'NoneType' object has no attribute 'rsplit'` has been fixed by #38 
 * Converting after patching with #38 led to `AttributeError: 'list' object has no attribute 'rsplit'`, which was fixed in #43 
 * Now the error is `RuntimeError: packed.cwl#workflow_prepare_fasta_db.cwl has no corresponding action`

**Comment by simleo** [comment]
> Now the error is `RuntimeError: packed.cwl#workflow_prepare_fasta_db.cwl has no corresponding action`

Fixed by #45. Now the RO bundle can be converted with no errors.
